### ISSUE TITLE
'DPM86: error 2' with $_cpu_vm_dpmi = "emulated"

### ISSUE THREAD

**Issue by magiblot** [dosemu2/dosemu2#1439]
**Describe the bug**
Setting the option `$_cpu_vm_dpmi = "emulated"` results in dosemu always failing on startup.

**To Reproduce**
Enable `$_cpu_vm_dpmi = "emulated"` in `~/.dosemurc` and then run dosemu.

Alternatively, have another VM application running in the background (e.g. VirtualBox) and then try to run dosemu with `$_cpu_vm_dpmi = "auto"`. It looks like dosemu is unable to use "kvm" in this circumstance and falls back on "emulated", even if the user never asked for it, thus failing anyway.

**Attach the log**
`-Da` log with `$_cpu_vm_dpmi = "emulated"`: [boot_2.zip](https://github.com/dosemu2/dosemu2/files/6218059/boot_2.zip)

**Additional info**
Using https://github.com/dosemu2/dosemu2/commit/32bd2894bdc3676d111574b38a21c2214c4b98cc, https://github.com/dosemu2/fdpp/commit/ad970d626ed33578d8b188df3aa4d8b8bc80ec9f and https://github.com/dosemu2/comcom32/commit/d64f704d370e75b2fdc48c247cdea14c34e8a754.

**Comment by stsp** [comment]
Even with freecom?

**Comment by stsp** [comment]
Where have you got comcom32?
From what package?

**Comment by magiblot** [comment]
Everything is from the AUR, so I compiled the packages myself. The packages refer to stable releases but their PKGBUILD has an option to use the latest git version.

https://aur.archlinux.org/packages/fdpp/
https://aur.archlinux.org/packages/comcom32/

> Even with freecom?

That's what I'm trying to figure out. Do I have to build dosemu without fdpp support in order to use freecom? If not, what's the simplest way to switch to freecom?

**Comment by stsp** [comment]
Just put it to C: and make a symlink:
command.com.

Please upload your comcom32 so
that I can try it myself.

**Comment by magiblot** [comment]
Thank you! That was simpler than I expected.

I can confirm that dosemu does not fail on startup with freecom.

Here is my comcom32, built with `i686-pc-msdosdjgpp-gcc (GCC) 9.3.0`: [comcom32.zip](https://github.com/dosemu2/dosemu2/files/6218118/comcom32.zip)

**Comment by stsp** [comment]
Yes, I confirm the problem.
Siiigh. :(
Bart, how difficult would it be to
implement CMOVcc family?

**Comment by stsp** [comment]
Can you try this patch as a work-around:
```
diff --git a/Makefile b/Makefile
index 2f5b1f0..926f708 100644
--- a/Makefile
+++ b/Makefile
@@ -6,7 +6,7 @@ DOS_LD ?= i586-pc-msdosdjgpp-gcc
 DOS_STRIP ?= i586-pc-msdosdjgpp-strip
 PREFIX ?= /usr/local
 DATADIR ?= $(PREFIX)/share/comcom32
-C_OPT = -Wall -O2 -finline-functions -Wmissing-declarations
+C_OPT = -Wall -O2 -finline-functions -Wmissing-declarations -march=i386
 LINK_OPT =
 OBJS = command.o cmdbuf.o version.o
 CMD = comcom32.exe
@@ -39,7 +39,7 @@ $(CMD): $(OBJS)
 
 # Common rules
 %.o : %.c %.h
-       $(REDIR) $(DOS_CC) $(C_OPT) $(C_OUTPUT) -c $< -o $@
+       $(DOS_CC) $(C_OPT) $(C_OUTPUT) -c $< -o $@
 
 ifeq (,$(wildcard $(CMD)))
 install:
```

**Comment by magiblot** [comment]
Yes, this solves the issue, and it also seems reasonable since DOS software does not usually target i686.

Here's the new binary if you want to try it: [comcom32-i386.zip](https://github.com/dosemu2/dosemu2/files/6218209/comcom32-i386.zip)

**Comment by stsp** [comment]
Applied, but its a quite dubios
fix. We need to support djgpp
fully because there are those
djgpp ports of some unix software,
not only comcom32.
I wouldn't be surprised if mugen
or some allegro-based games
would be compiled w/o -march=i386.

**Comment by Torinde** [comment]
Please note that the [CPUID bit](https://en.wikipedia.org/wiki/CPUID#EAX=1:_Processor_Info_and_Feature_Bits) for CMOV also means that FCMOV, F(U)COMI are supported if the x87 bit shows x87 FPU is present.

So, since dosemu2 already has x87, if you implement CMOV, then to avoid issues probably it's good to implement also FCMOV/FCOMI.
- [ ] CMOV
- [ ] FCMOV
- [ ] FCOMI

To complete Pentium Pro remain:
- [ ] RDPMC
- [ ] NOPL
- [ ] PAE

To complete Pentium remain (unless you already emulate those? I can't find them in the code):
- [ ] RDMSR / WRMSR - Model-specific registers
- [ ] RSM - System Management Mode
